File: Data Structures & Algorithms/surrounded-regions/submission-4.cpp
class Solution {
public:
    void solve(vector<vector<char>>& board) {
        if (board.size() == 0 or board[0].size() == 0) {
            return;
        }

        int ROWS = board.size();
        int COLS = board[0].size();

        for (int row = 0; row < ROWS; ++row) {
            for (int col = 0; col < COLS; ++col) {
                if (row == 0 || col == 0 || row == ROWS - 1 || col == COLS - 1) {
                    if (board[row][col] == 'O') {
                    dfs(board, row, col);
                    }
                }
            }
        }

        for (int row = 0; row < ROWS; ++row) {
            for (int col = 0; col < COLS; ++col) {
                if (board[row][col] == 'O') {
                    board[row][col] = 'X';
                } else if (board[row][col] == '#') {
                    board[row][col] = 'O';
                }
            }
        }
    }

private:
    void dfs(vector<vector<char>>& board, int row, int col) {
        if (row < 0 || col < 0 || row >= board.size() || col >= board[0].size() || board[row][col] != 'O'){
            return;
        }

        board[row][col] = '#';
        dfs(board, row + 1, col);
        dfs(board, row - 1, col);
        dfs(board, row, col + 1);
        dfs(board, row, col - 1);

    }
};

File: Data Structures & Algorithms/unique-paths-ii/submission-4.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        ROWS = len(obstacleGrid)
        COLS = len(obstacleGrid[0])

        if obstacleGrid[0][0] == 1:
            return 0 

        dp = [[0] * COLS for _ in range(ROWS)]
        dp[0][0] = 1

        for row in range(ROWS):
            for col in range(COLS):
                if row == 0 and col == 0:
                    continue
                if obstacleGrid[row][col] == 1:
                    dp[row][col] = 0
                    continue

                top = dp[row - 1][col] if row > 0 else 0
                left = dp[row][col - 1] if col > 0 else 0

                dp[row][col] = top + left;

        return dp[ROWS - 1][COLS - 1]
        